src/vm/costs/mod.rs

```diff
@@ -194,6 +260,235 @@ pub enum CostErrors {
     CostContractLoadFailure,
 }
 
+fn load_state_summary(clarity_db: &mut ClarityDatabase) -> Result<CostStateSummary> {
+    let last_processed_at = match clarity_db.get_value(
+        "vm-costs::last-processed-at-height",
+        &TypeSignature::UIntType,
+    ) {
+        Some(v) => u32::try_from(v.expect_u128()).expect("Block height overflowed u32"),
+        None => return Ok(CostStateSummary::empty()),
+    };
+
+    let metadata_result = clarity_db
+        .fetch_metadata_manual::<String>(
+            last_processed_at,
+            &STACKS_BOOT_COST_VOTE_CONTRACT,
+            "state_summary",
+        )
+        .map_err(|e| CostErrors::CostComputationFailed(e.to_string()))?;
+    let serialized: SerializedCostStateSummary = match metadata_result {
+        Some(serialized) => serde_json::from_str(&serialized).unwrap(),
+        None => return Ok(CostStateSummary::empty()),
+    };
+    Ok(CostStateSummary::from(serialized))
+}
+
+fn store_state_summary(
+    clarity_db: &mut ClarityDatabase,
+    to_store: &CostStateSummary,
+) -> Result<()> {
+    let block_height = clarity_db.get_current_block_height();
+
+    clarity_db.put(
+        "vm-costs::last-processed-at-height",
+        &Value::UInt(block_height as u128),
+    );
+    let serialized_summary =
+        serde_json::to_string(&SerializedCostStateSummary::from(to_store.clone()))
+            .expect("BUG: failure to serialize cost state summary struct");
+    clarity_db.set_metadata(
+        &STACKS_BOOT_COST_VOTE_CONTRACT,
+        "state_summary",
+        &serialized_summary,
+    );
+
+    Ok(())
+}
+
+///
+/// This method loads a cost state summary structure from the currently open stacks chain tip
+///   In doing so, it reads from the cost-voting contract to find any newly confirmed proposals,
+///    checks those proposals for validity, and then applies those changes to the cached set
+///    of cost functions.
+///
+fn load_cost_functions(clarity_db: &mut ClarityDatabase) -> Result<CostStateSummary> {
+    let last_processed_count = clarity_db
+        .get_value("vm-costs::last_processed_count", &TypeSignature::UIntType)
+        .unwrap_or(Value::UInt(0))
+        .expect_u128();
+    let cost_voting_contract = &STACKS_BOOT_COST_VOTE_CONTRACT;
+    let confirmed_proposals_count = clarity_db
+        .lookup_variable(&cost_voting_contract, "confirmed-proposal-count")
+        .map_err(|e| CostErrors::CostComputationFailed(e.to_string()))?
+        .expect_u128();
+    debug!("Check cost voting contract";
+           "confirmed_proposal_count" => confirmed_proposals_count,
+           "last_processed_count" => last_processed_count);
+
+    // we need to process any confirmed proposals in the range [fetch-start, fetch-end)
+    let (fetch_start, fetch_end) = (last_processed_count, confirmed_proposals_count);
```

Comment:
I'm not sure this is getting adequate stress-testing.  What happens if, for example, 10 different cost-functions all get approved in the same block?  What happens if all cost-functions are for the same target contract -- which one gets applied for interception?  Can we confirm that cost functions take effect only on the next _block_ (but not on the next _transaction_)?

Reply:
> What happens if, for example, 10 different cost-functions all get approved in the same block?

They would all be applied. I think we should probably add something like a minimum STX holding period for confirmed proposals, which would make confirming many cost-functions in the same block very difficult (and impose a limit on the number approved in a single block: the limit would be ~5 if the vote threshold is 20% of liquid stx).

> What happens if all cost-functions are for the same target contract -- which one gets applied for interception?

These proposals are applied in order, and proposals can override prior proposals.

> Can we confirm that cost functions take effect only on the next block (but not on the next transaction)?

The integration tests for this do test this behavior, but there isn't an explicit assertion in the code for this -- it's a consequence of the fact that the LimitedCostTracker is only constructed at the beginning of the block. The LimitedCostTracker could add an assertion that the block has not yet started.

Reply:
TODO: add a `panic!()` if the `LimitedCostTracker` is instantiated in the middle of a block

Reply:
> They would all be applied. I think we should probably add something like a minimum STX holding period for confirmed proposals, which would make confirming many cost-functions in the same block very difficult (and impose a limit on the number approved in a single block: the limit would be ~5 if the vote threshold is 20% of liquid stx).

I think this is a good idea, and something we should test to make sure a bunch of (expensive) proposals don't all hit the network at once.

Reply:
Added a panicking assertion in 196c893 that checks to make sure that there's no pending writes on the block.

Reply:
And, in 950d829, I added a check / error type to the contract that prevents more than 10 proposals being confirmed in the same block (and a test case for that limit). We can revisit that constant or how we're limiting it later on.

Reply:
Thanks!